content/posts/2023-09-19-Diagnostic-Improvements-in-Clang-17.md

```diff
@@ -0,0 +1,522 @@
+---
+author: "Takuya Shimizu"
+date: "2023-09-19"
+tags: ["clang", "frontend", "gsoc"]
+title: "Diagnostic Improvements in Clang 17"
+---
+
+## Introduction
+In the last few months, I have been a part of an ongoing effort to improve Clang's diagnostic capabilities.
+The newly released Clang 17 brings several of these improvements to the forefront.
+This blog post aims to provide a comprehensive overview of these diagnostic enhancements.
+We will employ simplified code examples and compare diagnostic outputs from Clang 16 and Clang 17 to illustrate how the latest updates can enhance the development experience for Clang users.
+
+### Multi-line printing of code snippet
+One of the most anticipated diagnostic features of Clang 17 is its support for multi-line printing of code snippets.
+This marks a departure from the old single-line limit, which used to make it difficult to fully understand the context around a code issue.
+This new feature improves the readability and comprehensibility of diagnostic messages by displaying a more complete view of the code in question.
+Moreover, line numbers are now attached to the left of each line, allowing for quicker navigation and issue resolution.
+
+```c++
+int func(
+  int a, int b, int& r);
+
+void test(int *ptr) {
+  func(3, 4, 5);
+  func(3, 4);
+}
+```
+
+Before:
+```console
+<source>:5:3: error: no matching function for call to 'func'
+  func(3, 4, 5);
+  ^~~~
+<source>:1:5: note: candidate function not viable: expects an lvalue for 3rd argument
+int func(
+    ^
+<source>:6:3: error: no matching function for call to 'func'
+  func(3, 4);
+  ^~~~
+<source>:1:5: note: candidate function not viable: requires 3 arguments, but 2 were provided
+int func(
+    ^
+```
+
+After:
+```console
+<source>:5:3: error: no matching function for call to 'func'
+    5 |   func(3, 4, 5);
+      |   ^~~~
+<source>:1:5: note: candidate function not viable: expects an lvalue for 3rd argument
+    1 | int func(
+      |     ^
+    2 |   int a, int b, int& r);
+      |                 ~~~~~~
+<source>:6:3: error: no matching function for call to 'func'
+    6 |   func(3, 4);
+      |   ^~~~
+<source>:1:5: note: candidate function not viable: requires 3 arguments, but 2 were provided
+    1 | int func(
+      |     ^
+    2 |   int a, int b, int& r);
+      |   ~~~~~~~~~~~~~~~~~~~~
+```
+
+In this example, the newly-covered source ranges make it easier to understand why the overload candidate is invalid.
+
+
+### Preprocessor-related diagnostics
+- Clang warns on macro redefinitins. When the redefinition happens in assembly files and the previous definition of the macro comes from the command line, the previous definition is diagnosed as coming from `<command line>` instead of `<built-in>`.
```

Comment:
```suggestion
- Clang warns on macro redefinition. When the redefinition happens in assembly files and the previous definition of the macro comes from the command line, the previous definition is now diagnosed as coming from `<command line>` instead of `<built-in>`.
```